connector/connect/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectStreamHandler.scala

```diff
@@ -117,7 +126,70 @@ class SparkConnectStreamHandler(responseObserver: StreamObserver[Response]) exte
       responseObserver.onNext(response.build())
     }
 
-    responseObserver.onNext(sendMetricsToResponse(clientId, rows))
+    responseObserver.onNext(sendMetricsToResponse(clientId, dataframe))
+    responseObserver.onCompleted()
+  }
+
+  def processRowsAsArrowBatches(clientId: String, dataframe: DataFrame): Unit = {
+    val spark = dataframe.sparkSession
+    val schema = dataframe.schema
+    // TODO: control the batch size instead of max records
+    val maxRecordsPerBatch = spark.sessionState.conf.arrowMaxRecordsPerBatch
+    val timeZoneId = spark.sessionState.conf.sessionLocalTimeZone
+
+    val rows = dataframe.queryExecution.executedPlan.execute()
+    var numBatches = 0L
+
+    if (rows.getNumPartitions > 0) {
+      val batches = rows.mapPartitionsInternal { iter =>
+        ArrowConverters
+          .toArrowBatchIterator(iter, schema, maxRecordsPerBatch, timeZoneId)
+      }
+
+      val obj = new Object
+
+      val processPartition = (iter: Iterator[(Array[Byte], Long, Long)]) => iter.toArray
```

Comment:
This breaks sorted results. A higher partition can complete earlier than lower ones thus breaking the order. That is why I the snippet I posted buffered the partitions in the handler, while the main thread scanned over them 1 by 1.

Reply:
with batch_id, we can send higher partition before lower ones